src/S3/MultipartCopy.php

```diff
@@ -127,16 +130,24 @@ private function createPart($partNumber, $partsCount)
         foreach ($params as $k => $v) {
             $data[$k] = $v;
         }
-
-        list($bucket, $key) = explode('/', ltrim($this->source, '/'), 2);
-        $data['CopySource'] = '/' . $bucket . '/' . implode(
-            '/',
-            array_map(
-                'urlencode',
-                explode('/', rawurldecode($key))
-            )
-        );
+        // The source parameter here is usually a string, but can be overloaded as an array
+        // if the key contains a '?' character to specify where the query parameters start
+        if (is_array($this->source)) {
+            $key = str_replace('%2F', '/', rawurlencode($this->source['source_key']));
+            $data['CopySource'] = '/' . $this->source['source_bucket'] . '/' . $key;
+        } else {
+            list($bucket, $key) = explode('/', ltrim($this->source, '/'), 2);
+            $data['CopySource'] = '/' . $bucket . '/' . rawurlencode(
+                    implode(
+                        '/',
+                        array_map('urlencode', explode('/', $key))
```

Comment:
The old code for this was calling `rawurldecode` on the `$key`, is the difference intentional?

Reply:
Just changed that code entirely.  The idea behind it is to turn go through the string encoding everything but slashes.  Since there's a much cleaner way to do that (which is also used elsewhere in the SDK), I moved around the logic for consistency and code cleanup.    

Tests are all still passing and all of the manual testing I did showed no change.  rawurldecode is now just called in the getInputSource method instead